node: cpumgr: fix allocation of full physical CPUs when the policy option `full-pcpus-only` is enabled

ffromani opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Make sure we allocate full cores when the `full-pcpus-only`  option is enabled, the existing pre-check is not sufficient.
We need to make sure we never get as far as allocating single CPUs, so the smallest unit we allocate is a virtual thread group.


#### Which issue(s) this PR fixes:
Fixes #113537 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Fix misallocation of CPUs when using the `full-pcpus-only` CPUManager option
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
NONE
```

ffromani committed:
node: cpumgr: extra test for full-pcpus-only

The CPUManager static policy can fail to deliver allocation
of full physical cores, and allocate pairs of unrelated virtual threads.

The root cause for the breakage is insufficient test coverage for the
`full-pcpus-only` option. We start fixing the coverage by adding the test
that first demonstrates the issue. Pending testcases review, we will
add more testcases in future commits.

Reported in issue https://github.com/kubernetes/kubernetes/issues/113537
Signed-off-by: Francesco Romani <fromani@redhat.com>

swatisehgal commented:
/triage accepted

fromanirh commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -477,6 +477,13 @@ func takeByTopologyNUMAPacked(topo *topology.CPUTopology, availableCPUs cpuset.C
 		return acc.result, nil
 	}
 
+	if cpuGroupSize > 1 {
+		return cpuset.NewCPUSet(), SMTAlignmentError{
+			RequestedCPUs: numCPUs,
+			CpusPerCore:   topo.CPUsPerCore(),
+		}
+	}
+
```
I'm still looking for the best approach. I'm not too fond of the current changes myself. And yes, we will need the check in both variants.

fromanirh commented:
/sig node

fromanirh commented:
/hold
In addition to the comments (which I fully agree with, the current changes could use some serious refinement), I want to make sure the test coverage is better. That was the real culprit back in time, and I want to improve in this area

fromanirh commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -416,15 +421,11 @@ func (p *staticPolicy) podGuaranteedCPUs(pod *v1.Pod) int {
-func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs int) (cpuset.CPUSet, error) {
+func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs, cpuGroupSize int) (cpuset.CPUSet, error) {
 	if p.options.DistributeCPUsAcrossNUMA {
-		cpuGroupSize := 1
-		if p.options.FullPhysicalCPUsOnly {
-			cpuGroupSize = p.topology.CPUsPerCore()
-		}
```
answered in https://github.com/kubernetes/kubernetes/pull/113924#discussion_r1024209933

fromanirh commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -343,6 +343,11 @@ func (p *staticPolicy) allocateCPUs(s state.State, numCPUs int, numaAffinity bit
 
 	allocatableCPUs := p.GetAvailableCPUs(s).Union(reusableCPUs)
 
+	cpuGroupSize := 1
+	if p.options.FullPhysicalCPUsOnly {
+		cpuGroupSize = p.topology.CPUsPerCore()
+	}
+
```
OTOH, honoring the flag also when reserving CPUs at policy creation time will also prevent people to reserve just 1 cpu automatically. Which is probably still legit, but also likely too big/surprising of a change

ffromani committed:
node: cpumgr: takeByTopology* return flag for aligned allocation

Make takeByTopology{Packed,Distributed} return a flag to signal
the caller if they managed to allocate the requested cpus in multiple
of cpuGroupSize unit. IOW, enable the functions to tell the caller
if the allocation is aligned; for extra clarity, here, by "aligned"
we mean "done in multiples of `cpuGroupSize`.

In order to report if the allocation is aligned, we need to push the
check down to the `cpuAllocator` functions. This way we can check
each substep of allocation. Despite being relatively large, the change
is also straightforward.

This flag will be used in the upcoming commits by the static policy to
enforce the `full-pcpus-only` option and verify the allocation.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: cpumgr: verify actual allocation when full-pcpus-only is requested

To make sure we allocate full cores when the `full-pcpus-only`
option is enabled, the pre check is not sufficient.
We need to make sure the actual allocation is made using only multiples
of the `cpuGroupSize` (usually a hyperthread pair).
To do so, we leverage the extra flag returned by the
takeByTopology{packed,Distributed} functions.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
note for reviewers: I'm now much happier with the overall direction, but the PR is still WIP.

klueska commented:
> note for reviewers: I'm now much happier with the overall direction, but the PR is still WIP.

Thanks. Just ping me when it's read for review.

fromanirh commented:
/hold cancel
@klueska the PR is now ready for review. In general I'd like to add a bit more unit-tests, still thinking about this.

klueska commented:
I took a quick look at this patch. My first impression is that it seems unnecessarily invasive to plumb all of this `aligned` information through every call while in the process of doing the actual allocation. Can you elaborate on where / how the issue you were seeing manifested and why doing some sort of pre-check isn't sufficient (even if it means doing a more sophisticated pre-check than what we are doing today).

bart0sh commented:
@fromanirh
> alternative implementation with stricter pre-check here: https://github.com/kubernetes/kubernetes/pull/114114

Should this one be closed in favour of  https://github.com/kubernetes/kubernetes/pull/114114 or you want to keep both PRs open?

fromanirh commented:
> @fromanirh
> 
> > alternative implementation with stricter pre-check here: #114114
> 
> Should this one be closed in favour of #114114 or you want to keep both PRs open?

Initially I planned to keep both open, but now I'm inclined towards the latter approach. Will reopen this one if we decide the new approach is not good enough (and no other alternatives emerge).
